byterun/win32.c

```diff
@@ -111,16 +193,23 @@ int caml_write_fd(int fd, int flags, void * buf, int n)
   int retcode;
   if ((flags & CHANNEL_FLAG_FROM_SOCKET) == 0) {
 #if defined(NATIVE_CODE) && defined(WITH_SPACETIME)
-  if (flags & CHANNEL_FLAG_BLOCKING_WRITE) {
-    retcode = write(fd, buf, n);
-  } else {
+    if (!(flags & CHANNEL_FLAG_BLOCKING_WRITE))
 #endif
     caml_enter_blocking_section();
-    retcode = write(fd, buf, n);
-    caml_leave_blocking_section();
+#if WINDOWS_UNICODE
+    if ((flags & CHANNEL_FLAG_CONSOLE) && caml_win32_major < 10 && n > 0
+        && console_supports_unicode())
+#else
+    if ((flags & CHANNEL_FLAG_CONSOLE) && GetConsoleOutputCP() == CP_UTF8
```

Comment:
Same question here as for the case of `win32_utf8_read` above: shouldn't we be using the currently selected code page instead of handling just the case of `UTF_8`?

Reply:
This shim is only necessary if the output code page is UTF8 - otherwise the C `write` function works correctly. The unique problem when CP_UTF8 is selected is that `WriteConsole` (via `WriteFile`) returns the wrong answer to the `write` function causing the correct Unicode character to be written followed by some of the UTF-8 sequence characters themselves.